File: src/main/java/dev/kensa/context/TestContainerFactory.java
package dev.kensa.context;

import dev.kensa.Issue;
import dev.kensa.Notes;
import dev.kensa.state.TestInvocationData;
import dev.kensa.state.TestState;
import dev.kensa.util.ReflectionUtil;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.extension.ExtensionContext;

import java.lang.reflect.AnnotatedElement;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static dev.kensa.util.ReflectionUtil.getAnnotation;
import static java.util.Collections.emptyList;

public class TestContainerFactory {

    public TestContainer createFor(ExtensionContext context) {
        Class<?> testClass = context.getRequiredTestClass();

        return new TestContainer(
                testClass,
                context.getDisplayName(),
                invocationDataFor(testClass),
                notesFor(testClass),
                issuesFor(testClass)
        );
    }

    private Map<Method, TestInvocationData> invocationDataFor(Class<?> testClass) {
        return ReflectionUtil.testMethodsOf(testClass)
                             .map(this::createInvocationData)
                             .collect(
                                     LinkedHashMap::new,
                                     (m, i) -> m.put(i.testMethod(), i),
                                     Map::putAll
                             );
    }

    private TestInvocationData createInvocationData(Method method) {
        return new TestInvocationData(
                method,
                deriveDisplayNameFor(method),
                notesFor(method),
                issuesFor(method),
                initialStateFor(method)
        );
    }

    private TestState initialStateFor(Method element) {
        return getAnnotation(element, Disabled.class)
                .map(a -> TestState.Disabled)
                .orElse(TestState.NotExecuted);
    }

    private String deriveDisplayNameFor(Method element) {
        return getAnnotation(element, DisplayName.class)
                .map(DisplayName::value)
                // TODO: build a sentence from the method name
                .orElse(element.getName());
    }

    private String notesFor(AnnotatedElement element) {
        return getAnnotation(element, Notes.class)
                .map(Notes::value)
                .orElse(null);
    }

    private List<String> issuesFor(AnnotatedElement element) {
        return getAnnotation(element, Issue.class)
                .map(Issue::value)
                .map(Arrays::asList)
                .orElse(emptyList());
    }
}

File: src/main/java/dev/kensa/context/TestContainerFactory.kt
package dev.kensa.context

import dev.kensa.*
import dev.kensa.output.TestWriter
import dev.kensa.state.TestMethodContainer
import dev.kensa.state.TestState.*
import dev.kensa.util.*
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.extension.ExtensionContext
import java.lang.reflect.AnnotatedElement
import java.lang.reflect.Method

class TestContainerFactory {
    fun createFor(context: ExtensionContext, testWriter: TestWriter, commonBasePackage: String = ""): TestContainer =
        context.requiredTestClass.run {
            TestContainer(
                this,
                deriveDisplayName { simpleName.unCamel() },
                createMethodContainers(),
                notes(),
                issues(),
                deriveMinimumUniquePackageName(commonBasePackage),
                testWriter
            )
        }

    private fun Class<*>.createMethodContainers(): Map<Method, TestMethodContainer> =
        testMethods()
            .map { method: Method -> method.createMethodContainer(autoOpenTab()) }
            .associateByTo(LinkedHashMap()) { invocation: TestMethodContainer -> invocation.method }

    private fun Method.createMethodContainer(autoOpenTab: Tab): TestMethodContainer =
        TestMethodContainer(
            this,
            deriveDisplayName { normalisedPlatformName.unCamel() },
            notes(),
            issues(),
            initialState(),
            autoOpenTab(autoOpenTab)
        )

    private fun Method.initialState() = if (hasAnnotation<Disabled>()) Disabled else NotExecuted

    private fun AnnotatedElement.autoOpenTab(default: Tab? = null) : Tab = findAnnotation<AutoOpenTab>()?.value ?: default ?: Kensa.configuration.autoOpenTab

    private fun AnnotatedElement.deriveDisplayName(lazyDefault: () -> String) = findAnnotation<DisplayName>()?.value ?: lazyDefault()

    private fun AnnotatedElement.notes(): String? = findAnnotation<Notes>()?.value

    private fun AnnotatedElement.issues(): List<String> = findAnnotation<Issue>()?.value?.toList() ?: emptyList()

    private fun <T> Class<T>.deriveMinimumUniquePackageName(commonBasePackage: String): String =
        commonBasePackage
            .takeIf { it.isNotBlank() and packageName.startsWith(it) }
            ?.let { packageName.replaceFirst(it, "") }
            ?.removeLeadingDots()
            ?: packageName

    private fun String.removeLeadingDots() =
        replace(Regex("^\\.+"), "")
}